doc/holo-build.8.pod
=encoding UTF-8

=head1 NAME

holo-build - cross-distribution system package compiler

=head1 SYNOPSIS

holo-build [I<option>...] I<file>

holo-build B<--help|--version>

=head1 DESCRIPTION

Holo adds a few sprinkles on top of package management to make it suitable for
basic configuration management tasks. Its prime directive is to have all
configuration statically declared and defined within packages, which can be
installed to add new configuration, or uninstalled to remove configuration from
a system.

However, the tools for creating system packages are optimized towards compiling
applications for tarballs, and introduce needless complexity when you just want
to package up a few files and list some dependencies. holo-build provides a
simple, distribution-independent package description language and generates a
system package from such a description.

For example, the following package description will build a package that
installs and configures C<systemd-timesyncd> with a custom NTP server:

    [package]
    name     = "hologram-systemd-timesyncd"
    version  = "1.0"
    author   = "Jane Doe <jane.doe@example.org>"
    requires = ["systemd"]

    [[file]]
    path     = "/etc/systemd/timesyncd.conf.d/server.conf"
    content  = """
        [Time]
        NTP=ntp.someserver.local
    """

    [[symlink]]
    # as created by `systemctl enable systemd-timesyncd`
    path     = "/etc/systemd/system/sysinit.target.wants/systemd-timesyncd.service"
    target   = "/usr/lib/systemd/system/systemd-timesyncd.service"

    [[action]]
    on     = "setup"
    script = "systemctl daemon-reload && systemctl start systemd-timesyncd"

    [[action]]
    on     = "cleanup"
    script = "systemctl stop systemd-timesyncd"

The package description format is described below.

=head1 OPTIONS

The only positional argument I<file> is the file name from where the package
definition will be read. If no such argument is given, the package definition
is read from standard input instead.

=over 4

=item B<-o>, B<--output>=I<filename>

Write the resulting package to I<filename>. If not given, write the package into
the working directory using the naming convention for the corresponding output
package format. If I<filename> refers to a directory, write the package into this
directory using the naming convention for the corresponding output package format.
If the I<filename> is C<->, write the package to standard output.

=item B<-f>, B<--force>/B<--no-force>

By default, C<holo-build> will fail if the target file already exists. This
behavior protects against the user forgetting to increase the version when
editing the package description. With C<--force>, the target file will be
overwritten when it exists.

This switch has no effect when C<--output=-> or C<--suggest-filename> is in effect.

=item B<--format>=I<format>

Generate a package of the specified format, instead of the default package
format for the current distribution. Valid values are C<debian>, C<pacman> and
C<rpm>.

B<WARNING:> RPM generation is considered experimental because RPM is
underdocumented and a bizarrely baroque format to begin with.

=item B<--suggest-filename>

Do not generate a package. After reading and validating the package definition,
just print on standard output the suggested filename for this package. The
printed file name is the same one that will be used when C<--no-stdout> is in
effect. The format for package filenames follows the recommendations of the
targeted distribution:

    $ cat input.toml

    [package]
    name = "foo"
    version = "1.0"

    $ holo-build --suggest-filename --format=debian < input.toml
    foo_1.0-1_any.deb
    $ holo-build --suggest-filename --format=pacman < input.toml
    foo-1.0-1-any.pkg.tar.xz

This option can be used when auto-generating Makefiles, where the output
filename needs to be known before C<holo-build> runs (for purposes of dependency
resolution).

=item B<--help>

Print out usage information.

=item B<-V>, B<--version>

Print out holo-build's version string.

=back

=head2 Deprecated options

These switches will be removed in the next major version.

=over 4

=item B<--debian>

Use C<--format=debian> instead.

=item B<--pacman>

Use C<--format=pacman> instead.

=item B<--reproducible>

No effect. All packages are now built reproducibly.

=item B<--rpm>

Use C<--format=rpm> instead.

=item B<--stdout>

Use C<--output=-> instead.

=back

=head1 PACKAGE DESCRIPTION FORMAT

Package descriptions are written in L<the TOML format|https://github.com/toml-lang/toml>.

Only the C<[package]> section is required. All other sections (and all fields
not marked as required) are optional.

=head2 C<[package]> section

This section is required and contains global properties for the package.

    [package]
    name    = "example-package"
    version = "1.2.5"
    epoch   = 2
    release = 3
    description = "An example package"
    author      = "Jane Doe <jane.doe@example.org>"
    requires    = [ "other-package >= 2.0" ]
    provides    = [ "example-package-api = 1.2" ]
    conflicts   = [ "bloatware-package" ]
    replaces    = [ "sample-package" ]

=over 4

=item B<name> (string, required)

The package name. This string may never contain any slashes or newlines. The
output format may impose additional charset or size restrictions on the package
name. As a rule of thumb, the pattern C<[a-z0-9-]+> is a safe set of allowed
package names.

For C<--format=debian>, the package name may contain C<[a-z0-9+-.]>, but the
first character must be alphanumeric.

For C<--format=pacman>, the package name may contain C<[a-z0-9@._+-]>, but the
first character may not be a hyphen.

=item B<version> (string, required)

The package version. To ensure sanity, holo-build enforces a relatively strict
pattern. Version numbers must be numbers (C<0|[1-9][0-9]*>) chained together by
dots:

    version = "1.2.3"
    version = "0.10"
    version = "20151015"

=item B<alpha>, B<beta> (unsigned integer)

If one of these is non-zero, the package is a prerelease version. For instance,
the following corresponds to the full version C<1.2.0-beta.3>:

    version = "1.2.0"
    beta = 3

The exact rendition of the version string depends on the output format.

=item B<epoch> (unsigned integer)

An increase in the epoch (default: 0) can be used to force the package to be
seen as newer than any previous version with a lower epoch. This is used when
the version numbering scheme for a package changes, breaking normal version
comparison logic.

=item B<release> (unsigned integer)

The release number (default: 1) can be appended if the same package is rebuilt
multiple times without its contents changing, in order for the built packages
to be distinguishable from one another.

=item B<description> (string)

A description of the purpose and contents of this package.

=item B<author> (string, required for C<--format=debian>)

The name and mail address of the package author, in the form C<< Name <address> >>:

    [package]
    author = "Jane Doe <jane.doe@example.org>"

=item B<architecture> (string)

The target architecture of the package. The default (C<any>) is fine unless the
package contains compiled binaries. Valid values include:

    all any noarch
    i386 i686
    x86_64 amd64
    arm armel armv5tl
    armv6h armv6hl
    armhf armv7h armv7hl
    arm64 aarch64

The values on each line are synonymous to each other. The big variety of
architecture strings comes from the big variety of supported target
distributions, who all define architecture strings in their own way. holo-build
will map the input architecture string to the one that's appropriate for the
given target distribution.

=item B<requires> (array of strings)

A list of other packages that must be installed when this package is installed.
Some package managers call these I<dependencies> instead. When a specific
package version is required, a version test can be appended, using one of the
operators C<=>, C<< > >>, C<< >= >>, C<< < >> or C<< <= >>. If multiple version
tests are required, the same package can be stated multiple times.

    [package]
    # require any version of foo, and a 2.x version of bar
    requires = [ "foo", "bar >= 2.0", "bar < 3.0" ]

When the package contains any files below C</usr/share/holo/$PLUGIN_ID>, a
requirement

    requires = [ "holo-$PLUGIN_ID" ]

is implied automatically.

For C<--format=pacman> only, a special syntax is allowed to require complete
package groups (by giving the groupname with a C<group:> prefix), and to
exclude certain packages or package groups from this group requirement (by
prefixing the dependency with C<except:>). For example, to have the package
require all packages from the C<xorg> group, except for the C<xorg-drivers>
group and the C<xorg-docs> package:

    [package]
    requires = [
        # require all packages from the xorg group
        "group:xorg",
        # except for xorg-docs and all packages in the xorg-drivers group
        "except:xorg-docs",
        "except:group:xorg-drivers",
    ]

=item B<provides> (array of strings)

A list of other packages (or virtual packages) that the software provides the
features of. This means that this package can satisfy another package's
requirement for the provided package. If a specific version is provided, this
can be specified with the same syntax as for C<requires>.

    [package]
    name     = "rewrite-of-foo"
    provides = [ "foo = 2.1" ] # this package acts like foo-2.1

For C<--format=debian>, the provided version syntax is not allowed. C<provides>
may only contain plain package names.

=item B<conflicts> (array of strings)

A list of other packages that may not be installed when this package is
installed. Version tests can be added using the same syntax as for C<requires>.

For C<--format=pacman>, the same special syntax is allowed as for C<requires>;
see there for details.

=item B<replaces> (array of strings)

A list of obsolete packages that this package replaces. If this package is not
installed, but one of the obsolete packages is installed, a system upgrade will
result in the obsolete package being uninstalled, and this package being
installed as a replacement.

If this package can replace the obsolete package, but it shall not be replaced
automatically, don't use this; reference the obsolete package in both
C<provides> and C<conflicts> instead.

For C<--format=pacman>, the same special syntax is allowed as for C<requires>;
see there for details.

=item B<setupScript> (string, deprecated)

A shell script that will be executed (as root) when the package is installed or
updated, after the package files have been extracted to the file system.

If the package contains any files below F</usr/share/holo>, then C<holo apply>
will automatically run before this setup script is executed, so you can rely on
provisioned files, user accounts and groups to be available already.

B<DEPRECATED:> This key will be removed in the next major release. Use
C<[[action]]> sections instead.

    # deprecated
    [package]
    setupScript = "echo Hello World"

    # write instead
    [[action]]
    on     = "setup"
    script = "echo Hello World"

=item B<cleanupScript> (string, deprecated)

A shell script that will be executed (as root) when the package is being
uninstalled, after the package files have been removed from the file system.

If the package contains any files below F</usr/share/holo>, then C<holo apply>
will automatically run before this cleanup script is executed.

B<DEPRECATED:> This key will be removed in the next major release. Use
C<[[action]]> sections instead.

    # deprecated
    [package]
    cleanupScript = "echo Hello World"

    # write instead
    [[action]]
    on     = "cleanup"
    script = "echo Hello World"

=item B<definitionFile> (string, deprecated)

A path where C<[[user]]> and C<[[group]]> sections will be placed inside the
package. See the description of C<[[user]]> and C<[[group]]> sections below.
When omitted, the default value is
F</usr/share/holo/users-groups/${package_name}.toml>.

B<DEPRECATED:> This key can be specified for compatibility reasons, but its use
is discouraged, and it will be removed in the next major release.

=back

=head2 C<[[file]]> section

Each one of these sections define a file to be added to the package.

    [[file]]
    path    = "/etc/foo.conf"
    mode    = "0600"
    owner   = "foouser"
    group   = "foogroup"
    # alternative 1
    content = """
        content for foo.conf
        content for foo.conf
    """
    # alternative 2
    contentFrom = "input.txt"

=over 4

=item B<path> (string, required)

The path to this file. The path must be absolute and may not have a trailing
slash.

=item B<content>/B<contentFrom> (string, exactly one required)

If the C<content> field is given, it contains the content of this file.
Alternatively, C<contentFrom> may reference a file whose contents will be used.
This file must be present at package-build time; relative paths will be
interpreted relative to the directory of the input file (or, if the package
definition is presented on standard input, to the current working directory of
the C<holo-build> process).

If C<content> is given, it may not be empty. To create an empty file, you can
use C</dev/null> as a source:

    [[file]]
    path        = "/etc/empty-file.conf"
    contentFrom = "/dev/null"

=item B<raw> (boolean)

To aid readability, the C<content> field allows strings to have indentation
which will automatically be pruned. The following two sections are equivalent:

    [[file]]
    path    = "/etc/example.conf"
    content = """
    foo
        bar
    baz
    """

    [[file]]
    path    = "/etc/example.conf"
    content = """
        foo
            bar
        baz
    """

To disable this behavior and use leading whitespace verbatim, set the C<raw>
flag:

    [[file]]
    path    = "/etc/example.conf"
    raw     = true
    content = """
        foo
            bar
        baz
    """

=item B<mode> (string)

The mode bits for this file. Since TOML does not support octal number
literals, this field must be given as a string containing an octal number:

    mode = "0600" # rw-------
    mode = "0755" # rwxr-xr-x

=item B<owner>/B<group> (string or int)

The owner (or group) for this file. If this field contains an integer, it is
interpreted as the ID of the user or group. If this field contains a string, it
is interpreted as a user/group name.

Since user/group names cannot be mapped to IDs at package build time, specifying
a name will result in the file being packaged as belonging to user/group
C<root>, and the actual user/group will be applied at install time using
L<chown(1)> or L<chgrp(1)>.

B<BUG:> For C<--format=rpm>, specification of numeric IDs is broken. When
encountering a numeric user/group ID in a holo-build package, RPM will say
things like

    user 42 does not exist - using root
    group 42 does not exist - using root

and fall back to UID/GID 0. As a workaround, call L<chown(1)> or L<chgrp(1)>
from the package's setup script to fix the file ownership.

=back

=head2 C<[[directory]]> section

Each one of these sections define a directory to be added to the package.

    [[directory]]
    path = "/var/lib/foo"
    mode = "0600"
    owner = "foouser"
    group = "foogroup"

Note that directories are usually created automatically when files are placed in
them. A C<[[directory]]> section is only required to include an empty directory
in the package, or to assign non-standard permissions or specific ownership to a
directory.

=over 4

=item B<path> (string, required)

The path to this directory. The path must be absolute and may not have a
trailing slash.

=item B<mode>/B<owner>/B<group>

These are the same as for C<[[file]]> sections; see above.

=back

=head2 C<[[symlink]]> section

Each one of these sections define a symlink to be added to the package.

    [[symlink]]
    path   = "/etc/foo.conf"
    target = "bar.conf"

=over 4

=item B<path> (string, required)

The path to this directory. The path must be absolute and may not have a
trailing slash.

=item B<target>

The symlink target. Both relative and absolute targets are acceptable.

=back

=head2 C<[[action]]> section

Each one of these sections define an action that can be executed by the
package manager. For example:

    [[action]]
    on     = "setup"
    script = "echo Just Installed"

=over 4

=item B<on> (string, required)

This field defines when this action will be executed. The following values are
valid:

    on = "setup"   # run right after package is installed or upgraded
    on = "cleanup" # run right after package is removed

If there are multiple actions with the same C<on> value, they will be executed
in the order in which they are given in the package description.

=item B<script> (string, required)

This field contains a shell script that will be run (as root) when the action
is executed.

=back

=head2 C<[[user]]> and C<[[group]]> sections

These can be used to provision user accounts and groups when the package is
installed. For example:

    [package]
    name    = "foobar"
    version = "1.0"

    [[group]]
    name   = "foobargroup"
    system = true

    [[user]]
    name   = "foobaruser"
    uid    = 285
    group  = "foobargroup"

The user and group definitions are validated at package compilation time, and
written into F</usr/share/holo/users-groups/${package_name}.toml>. Because of
this file, a dependency on C<holo-users-groups> is implied and C<holo apply> is
executed in the setup and cleanup scripts. So the previous example is
functionally equivalent to:

    [package]
    name          = "foobar"
    version       = "1.0"
    depends       = ["holo-users-groups"]

    [[file]]
    path    = "/usr/share/holo/users-groups/foobar.toml"
    content = """
        [[group]]
        name   = "foobargroup"
        system = true

        [[user]]
        name   = "foobaruser"
        uid    = 285
        group  = "foobargroup"
    """

    [[action]]
    on     = "setup"
    script = "holo apply"

    [[action]]
    on     = "cleanup"
    script = "holo apply"

The actual syntax and semantics of C<[[user]]> and C<[[group]]> sections is
described in L<holo-users-groups(8)>.

=head1 SEE ALSO

L<holo(8)>

=head1 AUTHOR

Stefan Majewsky

Further documentation is available at the project homepage: L<https://holocm.org>

Please report any issues and feature requests at GitHub: L<https://github.com/holocm/holo-build/issues>

=cut


CHANGELOG
v1.5 (2017-02-20)
=================

Deprecation notices:

- The `--debian`, `--pacman` and `--rpm` options are deprecated. Use the new
  option `--format={debian,pacman,rpm}` instead.
- The `--stdout` option and its negation are deprecated. Use the new `--output`
  option instead.

New features:

- The invocation syntax was cleaned up. The new options `--format` and
  `--output` replace multiple other options which are now deprecated (see
  above).
- `holo-build` will complain when trying to overwrite an existing target file
  unless instructed to do so with the new `--force` option. No complaint will be
  issued if the target file is identical to what `holo-build` would've written.
- The much-outdated shell completion widgets were updated and now implement the
  full invocation syntax, except for deprecated options.

v1.4 (2016-09-13)
=================

New features:

- The new `package.architecture` field can be used to build packages
  with compiled binaries for selected x86 and ARM architectures.

Miscellaneous:

- Fix unit test failure in Go 1.7.

v1.3 (2016-06-07)
=================

Deprecation notices:

- The `package.setupScript` and `package.cleanupScript` keys are deprecated.
  Use the new `[[action]]` syntax instead.
- The `package.definitionFile` key is deprecated, without replacement. See
  below.
- The `--(no-)reproducible` options are deprecated. All packages produced by
  holo-build are now fully reproducible.

New features:

- Package definitions can now contain `[[action]]` sections. This syntax
  replaces `package.setupScript` and `package.cleanupScript`.

Changes:

- `package.definitionFile` is not needed anymore (and deprecated, see above).
  The definition file's name is now derived from the package name by default.
- Packages are now always built fully reproducibly, so the `--reproducible`
  option is not needed anymore (and deprecated, see above).

Miscellaneous:

- Strip binaries during build.

v1.2 (2016-02-25)
=================

New features:

- Add RPM generator. RPMs will now be built automatically when running on RHEL,
  Fedora, SLES, openSUSE or Mageia. RPM support is considered experimental
  because of the abysmal state of documentation for the RPM package format.

Bugfixes:

- Fix version detection when building an unpacked release tarball.

Miscellaneous:

- Use golangvend to simplify the management of library dependencies.
- Share validation logic between generators.

v1.1.1 (2016-01-30)
===================

Bugfixes:

- Fix a bug which caused symlinks in packages to be unpacked as empty regular
  files.

v1.1 (2016-01-30)
=================

Changes:

- Packages are now built in memory, without constructing a temporary directory
  tree in the filesystem. This also means that the fakeroot dependency is gone.

Bugfixes:

- Fix a syntax error in the zsh autocompletion.
- Fix various small errors in the manpage.

v1.0 (2015-12-04)
=================

No changes compared to the beta release.

v1.0-beta.1 (2015-12-03)
========================

Changes:

- holo-build is now informed about the new plugin structure of Holo. When a
  package includes files below /usr/share/holo, it will now add automatic
  dependencies on the Holo plugins used, rather than on Holo itself. (The
  dependency on Holo is expected to be implied by the plugin package.)



src/holo-build.sh
#!/bin/sh
#
# Copyright 2015 Stefan Majewsky <majewsky@gmx.net>
#
# This file is part of Holo.
#
# Holo is free software: you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation, either version 3 of the License, or (at your option) any later
# version.
#
# Holo is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# Holo. If not, see <http://www.gnu.org/licenses/>.
#

# if a package format was specified explicitly, skip distribution detection
# (can also shortcut if just asked for --help or --version)
for ARG in "$@"; do
    case $ARG in
        --format|--debian|--pacman|--rpm|--help|--version)
            exec /usr/lib/holo/holo-build "$@" ;;
        *) ;;
    esac
done

# check distribution and choose appropriate package format
[ -f /etc/os-release ] && . /etc/os-release || . /usr/lib/os-release
DIST_IDS="$(echo "$ID $ID_LIKE" | tr ' ' ',')"

case ",$DIST_IDS," in
    *,arch,*)   exec /usr/lib/holo/holo-build --format=pacman "$@" ;;
    *,debian,*) exec /usr/lib/holo/holo-build --format=debian "$@" ;;
    *,fedora,*) exec /usr/lib/holo/holo-build --format=rpm "$@" ;;
    *,mageia,*) exec /usr/lib/holo/holo-build --format=rpm "$@" ;;
    *,suse,*)   exec /usr/lib/holo/holo-build --format=rpm "$@" ;;
    *)
        echo "!! Running on an unrecognized distribution. Distribution IDs: $DIST_IDS" >&2
        echo ">> Please report this error at <https://github.com/holocm/holo-build/issues/new>" >&2
        echo ">> and include the contents of your /etc/os-release file." >&2
        exit 255
        ;;
esac
